### DISCUSSION TITLE
RFD: Variants on Virtuals

### DISCUSSION THREAD

**Discussion by alecbcs** [spack/spack#47927] [Ideas]
## Introduction
Virtual packages do not allow for variant to be passed to their providers. 

As such this requires package maintainers to have knowledge of each provider, writing the required provider variant dependencies by hand. For example, if a package depends on `zlib +shared` a maintainer must currently write the following,

```python
depends_on("zlib-api")
depends_on("zlib-ng +shared", when="^[virtuals=zlib-api] zlib-ng")
depends_on("zlib +shared", when="^[virtuals=zlib-api] zlib")
```

This currently makes it difficult to correctly package applications that depend on specific functionality from virtual packages.

## Possible Solutions
1. (From @alalazo and @tgamblin) Extend `provides()` to accept arbitrary spec conditions in addition to virtual name / versions, allowing providers to supply variants and other generic conditions on virtuals.

**Comment by alalazo** [comment]
Virtuals are annotated on edges, and define a contract between the parent (asking for something) and the child (providing something). The variant that might be added to virtuals don't need to be reflected in the provider. 

One example we talked about for the future is better support for language standards. Under that scenario, many language virtuals could define e.g. the `std` multi-valued variant, and a compiler should be able to do stuff like:
```python
provides("cxx std=14", when="@X.Y:")
provides("fortran std=2003", when="@A.B:")
```
There is no need for providers, like e.g. `gcc`, to define the `std` multivalued variant themselves. Indeed, in cases like this one it is unwanted since it will conflate values from different virtuals.

**Reply by tgamblin** [reply]
Also, propagation isn't the right mechanism here.  Propagation is for dependent-dependency relationships and propagates to a sub-DAG.  

Proposal 1 above would extend the meaning to propagate "through" a virtual, but as @alalazo said there is no need for that.  Packages really should just need to provide the virtual with variant.

The issue is really that we do not (currently) do anything with variants, arch, etc. on virtual specs provided by packages.  We only look at the version.  It would not IMO be hard to add support for variants or other generic spec conditions on virtuals so that we could fully support `provides`/`when`.

**Reply by alecbcs** [reply]
I don't understand the thumbs down, but I like the idea of building this into `provides()` such that a provider can supply a variant to a virtual. Using the word "propagation" wasn't a requirement to re-use variant propagation code, but instead to talk about the problem in abstract where we need to be able to forward on variant information from a virtual to a provider so that you don't have to write the mess that is currently required.

**Reply by tgamblin** [reply]
> Using the word "propagation" wasn't a requirement to re-use variant propagation code, but instead to talk about the problem in abstract where we need to be able to forward on variant information from a virtual to a provider so that you don't have to write the mess that is currently required.

The point is that we do not need extra syntax, like `++` (which we call propagation) for doing this, and dependents shouldn't really need to know that they are depending on a virtual.

We've got a general enough syntax/directive setup already -- you can already say `provides("foo +bar", when="+bar")`.  We just need to add some tests and update the concretizer to respect it.

**Reply by alalazo** [reply]
> Have virtual packages declare variants that must be specified by all providers

This might create issues, because:
- A provider becomes the fat interface of all the virtuals it provides, and that is not necessary
- Virtuals can step on each other if they use the same variant name with different meaning / values

**Reply by alecbcs** [reply]
I think we're going to go with the solution you and Todd suggested, it makes a lot of sense to me to extend the provides directive in the concretizer to allow for arbitrary spec conditions.
